src/code.c

```diff
@@ -3636,7 +3606,24 @@ static Int PreSave (
   return 0;
 }
 
+static void InitModuleState(void *data)
+{
+    STATE(OffsBodyCount) = 0;
+    STATE(LoopNesting) = 0;
+    STATE(LoopStackCount) = 0;
+    STATE(StackStat) = NewBag( T_BODY, 64*sizeof(Stat) );
+    STATE(StackExpr) = NewBag( T_BODY, 64*sizeof(Expr) );
 
+#ifdef HPCGAP
+    STATE(OffsBodyStack) = AllocateMemoryBlock(MAX_FUNC_EXPR_NESTING*sizeof(Stat));
+    STATE(LoopStack) = AllocateMemoryBlock(MAX_FUNC_EXPR_NESTING*sizeof(UInt));
+#else
+    static Stat MainOffsBodyStack[MAX_FUNC_EXPR_NESTING];
+    static UInt MainLoopStack[MAX_FUNC_EXPR_NESTING];
+    STATE(OffsBodyStack) = MainOffsBodyStack;
+    STATE(LoopStack) = MainLoopStack;
+#endif
```

Comment:
So, I thought about how to actually use the `StateSlots` resp. the `data` pointer. And realized that they are mostly useless, unless we still safe the offset resp. data pointer somewhere in `STATE`, because otherwise the module cannot access them later on outside of the constructor/destructor.

So right now, these are only useful for the very narrow usecase where it suffices to have access in the constructor/destructor. This module might be such a case: we could put the memory `OffsBodyStack` and  `LoopStack` point at into the `StateSlots`: i.e., pass `MAX_FUNC_EXPR_NESTING*(sizeof(Stat) + sizeof(UInt))` to `RegisterModuleState`, then replace the above `#ifdef/#else/#endif` by something like this (untested):
```
  STATE(OffsBodyStack) = data;
  STATE(LoopStack) = (UInt *)((Stat *)data)+ MAX_FUNC_EXPR_NESTING);
```

That said, I am not sure that's a good idea...

And in any case, I'd be more interested in moving other things into the `StateSlots`, in order to improve module encapsulation and saturation. But these then would have to be accessed outside the constructor/destructor. The best I can come up with for that is to put all module specific state into a separate struct `MyModuleState`; then add `MyModuleState *myModuleState` to `GAPState`, and finally in the constructor do something like `STATE(myModuleState) = (MyModuleState *)data;`. Then later on, code outside of the constructor can access the module specific state via `STATE(myModuleState)`. Of course this means there is an indirection there. At this point, we might as well use `AllocateMemoryBlock` to store the state elsewhere. TThe "only" advantage of putting it into `StateSlots` is that these may, if we are lucky, be in the same cacheline as the start of `GAPState`... this, of course, can be quite beneficial.

Anyway, none of this is meant to speak against (or for) this PR, I am just thinking out loud... :-)

Reply:
Oh, and of course, this module separation comes at a price: accessing `STATE(foo)` requires one indirection less than `STATE(myModuleState)->foo`. I am fine with that for e.g. the coder, were speed is not that crucial, but for other things, we may want to do benchmarks before moving them around.

Reply:
I introduced `StateSlots` because @rbehrends said that it is preferable to have ThreadLocal state inside the TLS structure for performance reasons. Otherwise one could just allocate the memory for a module-local state struct as you say.

Reply:
Sure, I know that @rbehrends says this, and he has implemented this already before, see the `Extra` member of `struct ThreadLocalStorage`, which was added for that same purpose (and which this PR perhaps should remove, together with `TLS_NUM_EXTRA`?)

But (a) I don't quite understand why it is better (is it "just" because of cache locality? Or is there some other factor at play?), (b) right now *nothing* uses this mechanism, and so I thought it prudent to think (explicitly, in a way that others can see and read) about how one might use it; but actually, I think it would be better if we actually *did* use it -- or at least had plans to use it in the near future. Otherwise, it'll end up like its predecessor: yet another clever idea that nobody uses, everybody forgets about, and years down the road somebody stumbles about and  wonder about...
So if we don't start using with this PR, then perhaps at least add a comment somewhere which states the purpose of `StateSlots`?

Reply:
`StateSlots` as it stands does not work at all for HPC-GAP.

I now remember why I passed the offset into `StateSlots` in the constructor at first: In HPC-GAP `STATE` refers to a state struct *per thread*, so a module has a state per thread.

Now, if we store a `handle` (as I called it) in the module, this can be constant after the constructor has been called, `STATE` would be accessing `(MyModuleState *)(&TLS->StateSlots[MyModuleHandle])`.

`TLS` is a global variable pointing to the thread local memory region for each thread, in GAP there is only one instance of this state.

I just tried implementing this approach for the functions module (in the process introducing `IncRecursionDepth`, `DecRecursionDepth`, `GetRecursionDepth`, and `SetRecursionDepth`), and it works. 

The other option is letting `RegisterModuleState` just register a pointer in the `State` struct which the constructor sets (or which is set to the return value of the constructor, say). This is vaguely what you describe above, but would break the locality that @rbehrends seems to care about.

I am reluctant to do more large-scale refactor concerning state, and then rewrite it another ten times.

I thought I had removed the `TLSExtra` stuff but apparently did not commit that.

Reply:
> I now remember why I passed the offset into StateSlots in the constructor at first: In HPC-GAP STATE refers to a state struct per thread, so a module has a state per thread.

Of course. And we call the constructor per thread, too, so each gets its own `data` pointer, relative to its `STATE`.

> Now, if we store a handle (as I called it) in the module

What do you mean with "in the module" ?

Anyway, what I discussed above stored the pointer inside the `STATE` -- the only place where it makes sense, as this pointer of course is thread local.

> I just tried implementing this approach [...]

Which approach, exactly?

> The other option is letting `RegisterModuleState` just register a pointer in the State struct which the constructor sets (or which is set to the return value of the constructor, say). 

I don't follow. Why would `RegisterModuleState` do that, instead of the module constructor? If `RegisterModuleState` does it, how would a module learn (and remember) where its data is being stored?

> This is vaguely what you describe above [...] 

Only very vaguely (as far as I was able to guess what "your" scheme is, you don't really describe it in detail). At the very least, it differs in the crucial point that now once again, we have to come up with a scheme to let the module figure out where its data (or at least the pointer to its data) is being stored; if the module stores that pointer into `STATE` on its own, this problem goes away. And of course a module which doesn't need any extra data would simply not store the pointer.

Let me spell out once again what I pondered about (and I should stress that I am not saying we should use that pattern, just that it seems to be the most convenient if all one is given are `StateSlots` as implemented here right now:

1. Move all data members used by a module into a new struct, say `MyModuleState`:
2. Replace those members inside of `GAPState` by a single pointer member `MyModuleState *myModuleState;`
3. Invoke `RegisterModuleState(sizeof(MyModuleState), InitModuleState, ???);`
4. In `InitModuleState` store this pointer in `STATE(myModuleState)`. To summarize:
```C
// in my_module.h
typedef struct {
    UInt foo;
} MyModuleState;

// in my_module.c
static void InitModuleState(void *data)
{
    STATE(myModuleState) = (MyModuleState *)data;
    STATE(myModuleState)->foo = 42;
}

StructInitInfo * InitInfoMyModule ( void )
{
    RegisterModuleState(sizeof(MyModuleState), InitModuleState, 0);
    return &module;
}
```
This way, the module can freely access its custom TLS data (and other modules can, too, if `MyModuleState` is indeed declared in the header file; but of course one can also keep the data private, by moving the typedef into the C file, and using an opaque struct pointer type for `myModuleState`.


> [...] but would break the locality that @rbehrends seems to care about.

Once again, I don't follow: which locality are we talking about here? How is it broken? How does storing a pointer to a subsection of the `TLS` data inside the `TLS` break locality? Or perhaps you didn't mean locality, but instead are concerned about the extra indirection (`STATE(myModuleState)->foo` vs `STATE(myModuleStateFOO)`?